### PR TITLE
C#: Log compiler and extractor diagnostics

### PR DESCRIPTION
**Pull request by calumgrant**

### CODE REVIEW

**Comment by pavgust** [comment]
This pull request **introduces 4 alerts** when merging 2dc5256ab6615ab81262b0922dc05dc162ed1829 into 86040575b1533b28adf284ecd1af7a9220d015cc - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-ed3a9f5293bc4b8ecd8ede4310f40ec8b4dfc627)

**new alerts:**

* 1 for Generic catch clause
* 1 for Redundant ToString\(\) call
* 1 for Missed opportunity to use Select
* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by calumgrant** [comment]
Performance report is [here](https://git.semmle.com/gist/calum/611861d6cd7c7a1cdf08c5f30175cfb1)

**Comment by pavgust** [comment]
This pull request **introduces 4 alerts** when merging 606e4ae346ea565c8a70d9657155f91f4f926642 into ed0ef36427a25004061ae417958658261d73740d - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-4d2de2d8a5ccce4dc496017797914b3b6f78c8c9)

**new alerts:**

* 1 for Generic catch clause
* 1 for Redundant ToString\(\) call
* 1 for Missed opportunity to use Select
* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 7121a10ce8608f95c660d65a7444eeb542a643e4 into 76caad0fb48b15b65ab7db8e6bb7e4a4b6b3ba0e - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-9e98e11d49612372d3ad11acd5d697a4a1e6a624)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 14a7e27a2c4cb00221317de7d03cb32dcc41fa24 into a4de82de0675fb3c4b4b19ebaebee74216c1fecc - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-f4be56e7a88fe9aa7449b4a277518e55942d294a)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: csharp/ql/src/semmle/code/csharp/commons/Compilation.qll [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  File getFileCompiled(int i) { compilation_compiling_files(this, i, result) }
+
+  /** Gets a source file compiled in this compilation. */
+  File getAFileCompiled() { result = getFileCompiled(_) }
+
+  /** Gets a diagnostic associated with this compilation. */
+  Diagnostic getADiagnostic() { result.getCompilation() = this }
+
+  /** Gets a performance metric for this compilation. */
+  float getMetric(int metric) { compilation_time(this, -1, metric, result) }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Compilation.qll]: make `private`?

File: csharp/ql/src/semmle/code/csharp/commons/Diagnostics.qll [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    extractor_messages(this, severity, origin, text, element, location, stackTrace)
+  }
+
+  /** Gets the severity of this message. */
+  int getSeverity() { result = severity }
+
+  /** Gets the name of the extractor that produced this message, for example, `C# extractor`. */
+  string getOrigin() { result = origin }
+
+  /** Gets the text of this diagnostic. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/commons/Diagnostics.qll]: diagnostic -> message

File: csharp/ql/src/semmlecode.csharp.dbscheme [hunk 3]
```diff
@@ -1,3 +1,146 @@
+
+/**
+ * An invocation of the compiler. Note that more than one file may be
+ * compiled per invocation. For example, this command compiles three
+ * source files:
+ *
+ *   gcc -c f1.c f2.c f3.c
```
Comment (hvitved) [inline csharp/ql/src/semmlecode.csharp.dbscheme]: I realize that this was copied from C++, but should we have a `csc` example instead?

Comment (hvitved) [reply csharp/ql/src/semmlecode.csharp.dbscheme]: This is a general thing for all the comments, they are very C++ specific.

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 683b4e2095fa7f08390e29e50f1a14f980657e93 into e88e5cf4d77c12d1ce2ff75b210835df4f9f7407 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-fcc04133df7d748780733362ef31f24fba65b838)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: csharp/extractor/Semmle.Extraction.CSharp/Extractor.cs [hunk 4]
```diff
@@ -62,6 +62,7 @@ public void Analysed(int item, int total, string source, string output, TimeSpan
         /// <returns><see cref="ExitCode"/></returns>
         public static ExitCode Run(string[] args)
         {
+            var stopwatch = new Stopwatch();
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction.CSharp/Extractor.cs]: Indeed!

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 132a3f9823090eb85995d472bd8581c3106eb0df into c674f54129d381039bba85dfec776380562a01d5 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-d569805ecbd35adac589bc3cdfd0d961a2ea3117)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 94a791c53bbe15831322589c4de79bbc09ad5920 into c674f54129d381039bba85dfec776380562a01d5 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-474a11490df49626a4ea7e0579a5dd5e5b710ff0)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging d5404ad737d990e6a5dd8ba72e890aae27b4190b into 8adbfdaae7d6c9b1a73de0716ac78e74c5e288c2 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-4d3c5ba3f34270b89ed7e7513d8681a052df0a98)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging f7072df26890a9bbc74dac69b6c698f990d4b936 into 13e04f459d97d055bae5c775d6849e4e77ccc116 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-995a9235b86cecc841cf69c81e799bed788ba42d)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging b0bb3a5f331e2f3d07f23cf8e1ccb8e76eebe922 into 9653fbd4f7aa0d315854855743bd5c27144fd7f8 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-d1a5bd5263965362686065307ad1f80184a639f8)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 27b2eea4660f865df39a7c063f332890cb58f6ee into bb67ac9ad1adfc2ce248751757f4bfb74f2d5803 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-91be1e400d8ccd1ed7b6985b9b80c321f3e58f74)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging 88b57191039464113bc3d325def7db71d1a7f137 into bb67ac9ad1adfc2ce248751757f4bfb74f2d5803 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-8cc8fbdec9c684c45e8ad26a0a8181fef897334d)

**new alerts:**

* 1 for Generic catch clause

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**